crates/pallet-bridges-initializer-currency-swap/src/lib.rs

```diff
@@ -0,0 +1,397 @@
+        let mut weight = T::DbWeight::get().reads(1);
+
+        if current_native_bridge_balance == Zero::zero() {
+            let imbalance = T::NativeCurrency::withdraw(
+                &T::NativeTreasuryPot::get(),
+                amount,
+                frame_support::traits::WithdrawReasons::TRANSFER,
+                frame_support::traits::ExistenceRequirement::KeepAlive,
+            )?;
+            weight += T::DbWeight::get().writes(1);
+
+            T::NativeCurrency::resolve_creating(&T::NativeEvmBridgePot::get(), imbalance);
+            weight += T::DbWeight::get().writes(1);
+
+            return Ok(weight);
+        }
+
+        match current_native_bridge_balance.cmp(&amount) {
+            Ordering::Less => {
+                let imbalance = T::NativeCurrency::withdraw(
+                    &T::NativeTreasuryPot::get(),
+                    amount
+                        .checked_sub(&current_native_bridge_balance)
+                        .expect("current_native_bridge_balance is less than amount; qed."),
+                    frame_support::traits::WithdrawReasons::TRANSFER,
+                    frame_support::traits::ExistenceRequirement::KeepAlive,
+                )?;
+                weight += T::DbWeight::get().writes(1);
+
+                // We can safely ignore the result as overflow cann't be reached.
+                // current_native_bridge_balance < amount. The resulted balance is equal to amount.
+                let _ = T::NativeCurrency::resolve_into_existing(
+                    &T::NativeEvmBridgePot::get(),
+                    imbalance,
+                );
+                weight += T::DbWeight::get().writes(1);
+            }
+            Ordering::Greater => {
+                let imbalance = T::NativeCurrency::withdraw(
+                    &T::NativeEvmBridgePot::get(),
+                    current_native_bridge_balance
+                        .checked_sub(&amount)
+                        .expect("current_native_bridge_balance is greater than amount; qed."),
+                    frame_support::traits::WithdrawReasons::TRANSFER,
+                    frame_support::traits::ExistenceRequirement::KeepAlive,
+                )?;
+                weight += T::DbWeight::get().writes(1);
+
+                // We can safely ignore the result as overflow cann't be reached.
+                // current_native_bridge_balance + current_native_treasury < total_issuance.
+                // So, imbalance + current_native_treasury < total_issuance.
+                let _ = T::NativeCurrency::resolve_into_existing(
+                    &T::NativeTreasuryPot::get(),
+                    imbalance,
+                );
+                weight += T::DbWeight::get().writes(1);
+            }
+            Ordering::Equal => {}
+        }
+
+        Ok(weight)
+    }
+
+    /// Make evm bridge balance be provided amount value.
+    ///
+    /// The logic shouldn't change evm swappable balance value.
```

Comment:
```suggestion
    /// Make evm bridge balance be provided amount value.
    ///
    /// This function MINTS/BURNS the tokens as it needs to balance out the currencies and bridge pots.
    ///
    /// The logic shouldn't change evm swappable balance value.
```

Reply:
How about we add `debug_assert!`/test to this fn to make sure we don't alter the swappable balance?

Reply:
Thanks, good point - https://github.com/humanode-network/humanode/pull/716/commits/d10ea31094286f52c074feea4a12887b13b70629.